src/backend/distributed/operations/worker_copy_udf.c

```diff
@@ -0,0 +1,70 @@
+/*-------------------------------------------------------------------------
+ *
+ * worker_copy_udf.c
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_executor.h"
+#include "distributed/worker_shard_copy.h"
+
+PG_FUNCTION_INFO_V1(worker_copy);
+
+/*
+ * worker_copy copies a shard from this worker to another worker
+ *
+ * SQL signature:
+ *
+ * worker_copy(
+ *     source_table regclass,
+ *     target_node_id integer
+ *  ) RETURNS VOID
+ */
+Datum
+worker_copy(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint32_t targetNodeId = PG_GETARG_INT32(1);
+
+	if (IsCitusTable(relationId))
```

Comment:
Do we need this check? We're reading from the pg_dist_partition metadata, which we determined might not be ideal.

Reply:
After checking I guess we don't need it. Even without it we will get this error:
> +ERROR:  cannot execute a distributed query from a query on a shard
> +DETAIL:  Executing a distributed query in a function call that may be pushed to a remote node can lead to incorrect results.

Reply:
@marcocitus , @JelteF , by any chance, do you remember why reading form the pg_dist_partition metadata "might not be ideal" and why you relied on `ERROR: cannot execute a distributed query from a query on a shard`?
#6795 shows that in some cases that error may not occur.
May be you should reconsider this decision and check for Citus tables in `worker_copy_table_to_node`?

Reply:
I'm guessing because this might happen in the context of a transaction that modifies pg_dist_partition, and if those happen over a different connection then they might not be visible here.

Reply:
I'm not sure I understand what you mean. Could you be more specific about what problem might be caused by this check?

Basing on your comment I can only come up with the following:
There is a Citus table.
On connection A this table is converted to a regular table.
On connection B we call `worker_copy_table_to_node` for this table, and it fails with error for it doesn't see the changes and still sees that the table is a Citus table.

In this case why can't we wait until the changes become visible and retry `worker_copy_table_to_node`? It seems to me that till we see the table as a Citus table we shouldn't call `worker_copy_table_to_node` for it

Reply:
> In this case why can't we wait until the changes become visible 

because changes only become visible when the coordinator commits and finalizes a 2PC, which won't happen until it's done waiting for worker_copy_table_to_node, because committing is always the last thing it does